RFC 0001.md

```diff
@@ -0,0 +1,111 @@
+
+## Motivation and Scope
+Quoting [[1]], [[2]] and [[3]], the potential goals of this proposal are:
+
+1. Support subclassing `torch.Tensor` in Python
+2. Preserve `Tensor` subclasses when calling `torch` functions on them
+3. Preserve `Tensor` subclasses when calling `numpy` functions on them
+4. Use the NumPy API with PyTorch tensors (i.e. NumPy API calls dispatch to `torch` functions)
+5. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `Tensor` subclasses
+6. Reuse NumPy ufunc implementations directly from PyTorch
+7. Allow operations on mixed array types, e.g. `tensor + ndarray`
+8. Preserve `Tensor` subclasses when calling `Tensor` methods.
+9. Propagating subclass instances correctly also with operators, using views/slices/indexing/etc.
+10. Preserve subclass attributes when using methods or views/slices/indexing.
+11. A way to insert code that operates on both functions and methods uniformly (so we can write a single function that overrides all operators).
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have a much more streamlined experience. Namely, the following code example will work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors hooks to run whenever methods or operators are called, and to modify the result to their specific use-case, perform logging, or otherwise change the result or the action of the method.
+
+## Backwards Compatibility
+### With PyTorch `master` as of writing
+PyTorch `master` pointed to commit hash `957a07ffbd13d8a805f4d718e0282efc5d2bff85` at the time of writing. Any classes implementing `__torch_function__` based on the usage in this commit hash will break completely, due to the differing signature of the protocol. However, as a release hasn't been made with `__torch_function__` in it, this is a minor-impact issue.
+
+### With NumPy
+As we are using a different protocol compared to NumPy `__torch_function__` vs `__array_function__`, there is no difference to the usage for those using NumPy. We propose to delay the issue of allowing the usage of Torch tensors with NumPy functions to a separate RFC. 
```

Comment:
Sorry, I wasn't clear in my earlier comment. My question is not allowing PyTorch tensors be used with Numpy functions (although this is an interesting question to pose), but if I am a Numpy user who is familiar with the `__array_function__` API, and I come to PyTorch expecting `__torch_function__` to work the same way, will my expectations be surprised in any way?

Reply:
I don't think there are more than a few handful of users that have built up an intuition here since it's so new, nor do I think including methods will be particularly surprising.